canonical/pycloudlib
# Add continuous deployment with semantic versioning

Squash merge this one. The latest commit is to make it easier to see the untested PyPI parts.
I have already added my pycloudlib key to this repo (any committer can replace it (but not view it) by going into settings->secrets) so upload should work as it did to the test PyPI.

# Proposed Commit Message
```
Add continuous deployment with semantic versioning

Add new VERSION file to hold version. Enforce with Github Actions that
it must be updated on every PR. Also include Github Action to tag
and deploy to PyPI after PR is merged. Docs also updated accordingly.
```

# Test Procedure
I tested on my fork ensuring version checking, tagging, and uploading to the TEST PyPI works correctly.

ON A FORK...
checkout main
echo "0.0.1" > VERSION
commit and push

checkout this PR
git reset HEAD~ --hard (to checkout the commit that uses test.pypi.org rather than real PyPI)
change VERSION to "0.0.1"
commit and push to fork and PR against fork's main
`Check Semantic Versioning` check should fail with
`VERSION file must be updated`

change version to "1.2.3.4.5"
commit and push
`Check Semantic Versioning` check should fail with
`VERSION must be in 1!<major>.<minor>.<patch> format, not 1.2.3.4.5`

Remove latest commit so version goes back to 1.0.0 (git reset HEAD~ --hard)
commit and push
`Check Semantic Versioning` check should pass

Checkout main and git reset upstream/main --hard
commit and push to main
Retrigger PR checks (e.g., open and close)
`Check Semantic Versioning` check should pass

(Attempting to reproduce past this step won't work unless you're a maintainer of the pycloudlib project on test.pypi.org and you bump the version to something other than what already exists there. See https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/)
Merge PR to fork's main
Ensure repo now has a 1.0.0 tag
Ensure https://test.pypi.org/project/pycloudlib/#history has 1.0.0
